name: test
on: [push]

jobs:
  test:
    runs-on: ubuntu-18.04
    strategy:
      matrix:
        otp: ['21.x', '22.x', '23.x']
        elixir: ['1.8.x', '1.9.x', '1.10.x']
    services:
      dynamodb:
        image: amazon/dynamodb-local
        ports:
          - 8000:8000
    env:
      MIX_ENV: test
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-elixir@v1
        with:
          otp-version: ${{ matrix.otp }}
          elixir-version: ${{ matrix.elixir }}
      - name: cache mix.deps
        id: cache-deps
        uses: actions/cache@v2
        with:
          path: |
            deps
            _build
          key: v1-surge-${{ hashFiles(format('{0}{1}', github.workspace, '/mix.lock')) }}
          restore-keys: v1-surge-
      - name: execute mix deps.get if cache miss happened
        if: steps.cache-deps.outputs.cache-hit != 'true'
        run: |
          mix deps.get
          mix deps.compile
      - run: mix compile --warnings-as-errors
      - run: mix format --check-formatted
      - run: mix test
